packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,279 @@
+  const filteredActions = actions.filter((action) => {
+    let shouldExcludeCustomAction = false;
+    const eventType = action.eventData?.type || '';
+    const eventValue = action.eventData?.value || '';
+
+    if (eventType && eventValue) {
+      shouldExcludeCustomAction =
+        excludedCustomActions.includes(eventType) ||
+        excludedCustomActions.includes(eventValue);
+    }
+    return (
+      action.actionType !== UserActionType.CUSTOM || shouldExcludeCustomAction
+    );
+  });
+  return filteredActions;
+};
+
+export const mapUserActions = (rawActions: rawUserAction[]): UserAction[] => {
+  const mappedUserActions = rawActions.map((rawAction) => {
+    const actionData = JSON.parse(rawAction.value);
+    return {
+      actionType: rawAction.name as UserActionType,
+      timestamp: rawAction.time,
+      eventData: {
+        type: actionData.event_type,
+        value: actionData.event_value,
+      },
+      cause: actionData.cause,
+      searchHub: actionData.origin_level_1,
+      document: {
+        title: actionData.title,
+        uriHash: actionData.uri_hash,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdKey: actionData.c_contentidkey,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdValue: actionData.c_contentidvalue,
+      },
+      query: actionData.query_expression,
+    };
+  });
+  return mappedUserActions;
+};
+
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: string
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+    caseSessionFound: false,
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isPartOfTheSameSession(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+
+    handleSessionEnd(currentSession, returnTimeline, ticketCreationDate);
+
+    currentSession = {
+      start: action.timestamp,
+      end: action.timestamp,
+      actions: [action],
+    };
+  });
+
+  // Inserting ticket creation action in current session
+  const caseCreationSessionActions: UserAction[] | [] =
+    insertTicketCreationActionInSession(
+      returnTimeline.session,
+      ticketCreationDate
+    );
+
+  const caseCreationSession = createCaseCreationSession(
+    caseCreationSessionActions
+  );
+  returnTimeline.session = caseCreationSession;
+
+  return returnTimeline;
+};
+
+// Checks if an action is within 30mins of the previous action
+export const isPartOfTheSameSession = (
+  action: UserAction,
+  previousEndDateTime: string
+): boolean => {
+  return (
+    Math.abs(Number(action.timestamp) - Number(previousEndDateTime)) <
+    MAX_MSEC_IN_SESSION
+  );
+};
```

Comment:
One thing to note, to increase readability, create these small utility functions with the mindset to decouple any user agent logic  from the small tasks these utility functions are doing.
```suggestion
export const isActionWithin30Minutes = (
  action: UserAction,
 timestamp: string
): boolean => {
  return (
    Math.abs(Number(action.timestamp) - Number(timestamp)) <
    THIRTY_MINUTES
  );
};
```

if you wanna make it more general, you can even pass the interval as a parameter:

```
export const isActionWithinInterval = (
  action: UserAction,
 timestamp: string,
 interval = THIRTY_MINUTES
): boolean => {
  return (
    Math.abs(Number(action.timestamp) - Number(timestamp)) <
    interval
  );
};

```

Reply:
I like `isActionWithin30Minutes`, I dont think it needs to be general here since we only use it for the 30mins interval and this interval is unlikely to change since UA uses that as interval for sessions.

Reply:
Yup that name is really descriptive and I like it.